domain/src/main/java/com/thoughtworks/go/config/materials/git/GitMaterial.java

```diff
@@ -443,13 +314,136 @@ public GitMaterial withShallowClone(boolean value) {
+                        }
+                    }
+                });
+            }
+            int cloneDepth = shallowClone ? preferredCloneDepth : Integer.MAX_VALUE;
+            int returnValue;
+            if (executionContext.isServer()) {
+                returnValue = gitCommand.cloneWithNoCheckout(outputStreamConsumer, urlForCommandLine());
+            } else {
+                returnValue = gitCommand.clone(outputStreamConsumer, urlForCommandLine(), cloneDepth);
+            }
+            bombIfFailedToRunCommandLine(returnValue, "Failed to run git clone command");
+        }
+        return gitCommand;
+    }
+
+    private List<SecretString> secrets() {
+        SecretString secretSubstitution = line -> line.replace(urlForCommandLine(), getUriForDisplay());
+        return Collections.singletonList(secretSubstitution);
+    }
+
+    // Unshallow local repo to include a revision operating on via two step process:
+    // First try to fetch forward 100 level with "git fetch -depth 100". If revision still missing,
+    // unshallow the whole repo with "git fetch --2147483647".
+    private void unshallowIfNeeded(GitCommand gitCommand, ConsoleOutputStreamConsumer streamConsumer, Revision revision) {
+        if (gitCommand.isShallow() && !gitCommand.containsRevisionInBranch(revision)) {
+            gitCommand.unshallow(streamConsumer, UNSHALLOW_TRYOUT_STEP);
+
+            if (gitCommand.isShallow() && !gitCommand.containsRevisionInBranch(revision)) {
+                fullyUnshallow(gitCommand, streamConsumer);
+            }
+        }
+    }
+
+    private void fullyUnshallow(GitCommand gitCommand, ConsoleOutputStreamConsumer streamConsumer) {
+        if (gitCommand.isShallow()) {
+            gitCommand.unshallow(streamConsumer, Integer.MAX_VALUE);
+        }
+    }
+
+    private boolean isSubmoduleFolder() {
+        return getSubmoduleFolder() != null;
+    }
+
+    private boolean isGitRepository(File workingFolder) {
+        return new File(workingFolder, ".git").isDirectory();
+    }
+
+    private boolean isRepositoryChanged(GitCommand command, File workingDirectory) {
+        UrlArgument currentWorkingUrl = command.workingRepositoryUrl();
+        LOG.trace("Current repository url of [{}]: {}", workingDirectory, currentWorkingUrl);
+        LOG.trace("Target repository url: {}", url);
+        return !MaterialUrl.sameUrl(url.forDisplay(), currentWorkingUrl.forDisplay())
+                || !isRemoteFetchConfigEqual(command)
+                || !isBranchEqual(command)
+                || (!shallowClone && command.isShallow());
+    }
+
+    private boolean isRemoteFetchConfigEqual(GitCommand command) {
+        if (command.hasRefSpec()) {
+            try {
+                return ("+" + command.expandRefSpec()).equals(command.getConfigValue("remote.origin.fetch"));
+            } catch (Throwable ignored) {
+                return false;
+            }
+        }
+        return true;
+    }
```

Comment:
This is the last main bit; essentially, we should verify that the fetch configuration is equivalent as well when testing `isRepositoryChanged()`.